Show "Unavailable" and more descriptive tooltip when Google API fails

tomalec opened the pull request:
_Requires #745_

### Changes proposed in this Pull Request:

- Show "Unavailable" and a more descriptive tooltip, on the report pages, when Google API fails to respond with data. (41d2a0d)
	Fixes #748


### Screenshots:
![programs](https://user-images.githubusercontent.com/17435/121088257-bdc25580-c7e5-11eb-8c7e-8ab9dfa61ae4.gif)



### Detailed test instructions:

0. Reproduce scenario when Google does not respond with Free Listings report, or mock the data with 
	  ```php
	  add_filter(
	      'gla_prepared_response_mc-reports-programs',
	      function( $response ) {
		      return [];
	      },
	  );
	  add_filter(
	      'gla_prepared_response_ads-reports-programs',
	      function( $response ) {
		      return json_decode( file_get_contents( __DIR__ . '/tests/mocks/ads/reports/programs.json' ), true ) ?: [];
	      },
	  );
	  add_filter(
	      'gla_prepared_response_ads-reports-products',
	      function( $response ) {
		      return json_decode( file_get_contents( __DIR__ . '/tests/mocks/ads/reports/products.json' ), true ) ?: [];
	      },
	  );
	  add_filter(
	      'gla_prepared_response_mc-reports-products',
	      function( $response ) {
		      return [];
	      },
	  );
	  ```
1. Go to **GLA > Reports** 
2. Hover over tooltips in the performance summary section
3. Switch between `All Google programs, Single "Free", Single "Campaign"`
4. Switch between Programs and Products reports.


### Changelog entry

> Tweak - Show "Unavailable" and more descriptive tooltip when Google reports API fails

### Additional comments:
- For some reason, the link from the tooltip, does not work as `external` even though it's marked as one.
- [x] ~Currently, we show the same message for metrics that miss data due to API failure (`500`, no data), and when the API responds successfully, but without this particular metrics (the "Before Q1 2022" -scenario)~

### Doubts:
@j111q Do you have any comments for the below?

1. I made the `merchants.google.com` a clickable link, to make it easier for users to go there.
2. By default `Tooltip`'s text is aligned to the center, I [set it to `left`](https://github.com/woocommerce/google-listings-and-ads/commit/17e0610c0575be48e4ac9e39b66ff0e2caec7818) in this case, but maybe we should keep it consistent?
	![image](https://user-images.githubusercontent.com/17435/120866152-a5f19400-c58f-11eb-807a-9afb7271198c.png)

tomalec committed:
Show "Unavailable" and more descriptive tooltip,

on the report pages when Google API fails to respond with data.

Fixes https://github.com/woocommerce/google-listings-and-ads/issues/748

tomalec committed:
Make Metric `Tooltip`'s content align to left.

tomalec committed:
Swicth `missingFreeListingsData` to enum,

to distinguish the data missing due to API failing to return it vs. us not requesting it/Google API not supporting a given metric.

tomalec committed:
Allow Reports tooltip to wrap lines.

Addresses https://github.com/woocommerce/google-listings-and-ads/pull/749#issuecomment-855521668.
Set `min-` and `max-width` in case `@wordpress/components.Tooltip` would start to support both.  (Currently `min-width` = `width`)

ecgan reviewed (commented):
A few comments, two with ❓ . Briefly tested in my local machine, seems to be working ok.

tomalec commented on js/src/data/utils.js:
```diff
@@ -106,17 +130,19 @@ export function mapReportFieldsToPerformance( primary, secondary ) {
- * @param {boolean} [missingFreeListingsData] Flag indicating whether the data miss entries from Free Listings.
+ * @param {number} [value] The primary report fields fetched from report API.
+ * @param {number} [base] The secondary report fields fetched from report API.
+ * @param {MISSING_FREE_LISTINGS_DATA} [missingFreeListingsData] Flag indicating whether the data miss entries from Free Listings.
  * @return {PerformanceData} The calculated performance data of each metric.
  */
 export function fieldsToPerformance( value, base, missingFreeListingsData ) {
-	let delta = 0;
-
-	if ( value !== base ) {
-		const percent = ( ( value - base ) / base ) * 100;
-		delta = Number.isFinite( percent ) ? round( percent ) : null;
+	let delta = null;
+	if ( typeof value === 'number' && typeof base === 'number' ) {
+		delta = 0;
+		if ( value !== base ) {
+			const percent = ( ( value - base ) / base ) * 100;
+			delta = Number.isFinite( percent ) ? round( percent ) : null;
+		}
 	}
 
 	return { value, delta, prevValue: base, missingFreeListingsData };
```
I agree, that's pretty weird, that the majority of the code here is about the data calculation, while it's called `fieldsToPerformance`. Also, I don't like the fact we keep translating objects from `secondary` -> `base` -> `prevValue`.

Speaking of `missingFreeListingsData` we do need it for translation. And we don't need it for delta calculation, currently. But I think we may like to use it in future, when `SummaryNumber` accepts other `deltas`, to show for example "Unknown" or "**~** 5%" when there is some `missingFreeListingsData`.

WDYT of splitting it into two:
```js
export const fieldsToPerformance = ( value, base, missingFreeListingsData ) => ( {
	value,
	delta: calculateDelta( value, base ),
	prevValue: base,
	missingFreeListingsData
} );

function calculateDelta( value, base ) {
	let delta = null;
	if ( typeof value === 'number' && typeof base === 'number' ) {
		delta = 0;
		if ( value !== base ) {
			const percent = ( ( value - base ) / base ) * 100;
			delta = Number.isFinite( percent ) ? round( percent ) : null;
		}
	}

	return delta;
}
```

ecgan replied:
I think that looks much better. I just realized `fieldsToPerformance` is used in a few places, so with your suggestion here, `fieldsToPerformance` serves like a convenience conversion function to turn those parameters into an object with a specific schema, with `calculateDelta` responsible for doing the delta calculation. It's much easier to read. I like that. 🙂 👍

tomalec committed:
Call translation inline, fix the typo for "Unavailable" metric value.

Addresses https://github.com/woocommerce/google-listings-and-ads/pull/749#discussion_r647310125

tomalec committed:
Make the variable name more descriptive for `googleMCReportingDashboardURL`.

Addresses https://github.com/woocommerce/google-listings-and-ads/pull/749#discussion_r647311922

tomalec commented on js/src/reports/metric-number.js:
```diff
@@ -48,19 +53,52 @@ const MetricNumber = ( {
+			<>
+				{ createInterpolateElement(
+					__(
+						'This data is currently available for paid campaigns only.<br/><br/>Please try again later, or go to <link>merchants.google.com</link> to track your performance for Google Free Listings. ',
```
I wanted to give the translators the opportunity to divide the text into as many (or none) lines as they find it applicable for their copy. Is that an anti-pattern? Or maybe you meant something like below?
```js
__(
'<p>This data is currently available for paid campaigns only.<p><p>Please try again later, or go to <link>merchants.google.com</link> to track your performance for Google Free Listings.</p>',
'google-listings-and-ads'
),
```

ecgan replied:
I don't think the translators would care about those elements like `<p>` or `<br/>` or the UI; those elements might even be unnecessary noise to the translators. 

What I meant is something like this: 

```js
infoText = (
			<>
				<div>
					{ /* first sentence. */ }
					{ __(
						'This data is currently available for paid campaigns only.',
						'google-listings-and-ads'
					) }
				</div>
				<div>
					{ /* second sentence. */ }
					{ createInterpolateElement(
						__(
							'Please try again later, or go to <link>merchants.google.com</link> to track your performance for Google Free Listings.',
							'google-listings-and-ads'
						),
						{
							br: <br />,
							link: (
								<TrackableLink
									eventName="gla_google_mc_link_click"
									eventProps={ {
										context: 'reports',
										href: googleMCURL,
									} }
									type="external"
									target="_blank"
									href={ googleMCURL }
								/>
							),
						}
					) }
				</div>
			</>
		);
```

Note that I wrap the sentences with `div`s; it can be `<p>` or `<span>` depending on our UI needs, and it would be independent from the translators.

tomalec replied:
> What I meant is something like this:

My concern here was that we are hard-coding the fact there are two sentences (no less, no more). Which in my opinion was more of a copywriter/translator decision on how much text would end up there and how it would be divided into sentences. For example, somebody would like to phrase it more like:
> "Please try directly at merchants.google.com, as this data is currently not available for free listings"

And they would not be able to do that, without leaving the awkward empty space before or after.

But maybe, given my make-public-service-components background, I'm just mixing the roles of translator and copywriter here, as we already have specified copy for two separate messages.

tomalec replied:
Also, in my mental model, it was a single message: the explanation why the data may not be accurate.

tomalec replied:
Does [`979e77b` (#749)](https://github.com/woocommerce/google-listings-and-ads/pull/749/commits/979e77b590c59a8f81e0f799e34c2fe8fc4e9486) look better for you?

ecgan replied:
Yeah I think it does. 🙂

tomalec commented on js/src/reports/metric-number.js:
```diff
@@ -11,11 +11,16 @@ import GridiconInfoOutline from 'gridicons/dist/info-outline';
 const numberFormatSetting = { precision: 0 };
 
+const unavalable = __( 'Unavailable', 'google-listings-and-ads' );
+const googleMCURL = 'https://merchants.google.com/mc/reporting/dashboard';
```
I'll also rename it in the dashboard https://github.com/woocommerce/google-listings-and-ads/blob/add/748-no-mc-reports/js/src/dashboard/summary-section/performance-card.js#L16:L16

tomalec commented on js/src/reports/metric-number.js:
```diff
@@ -48,19 +53,52 @@ const MetricNumber = ( {
+						br: <br />,
+						link: (
+							<TrackableLink
+								eventName="gla_google_mc_link_click"
```
To me, it's already quite separate from other clicks. This track event name is only triggered 
- here
- [in the Dashboard](https://github.com/woocommerce/google-listings-and-ads/blob/8bc5c6aaa07763e6e48778be62675fd9769363b6/js/src/dashboard/summary-section/performance-card.js#L47)

in **both cases** the link is available in the same condition: Google free listings reporting API failed to respond with data. The only difference is the page from which it's fired (which is given in the `context` property).

Or maybe it's the lack of `dashboard` in the event name that is confusing here? (similarly to https://github.com/woocommerce/google-listings-and-ads/pull/749#discussion_r647311922)

ecgan replied:
> maybe it's the lack of `dashboard` in the event name that is confusing here?

D'oh you're right 🤦  I thought it was an existing generic Google MC link click to `https://merchants.google.com/`.

And to be more specific, I think it's more important to have `reporting` than `dashboard` in the event name. `dashboard` sounds like a generic dashboard page, but here we are directing users to _reporting dashboard_ because of "Google free listings _reporting API_ failed to respond with data".

tomalec replied:
>  I think it's more important to have reporting than dashboard in the event name. dashboard sounds like a generic dashboard page, but here we are directing users to reporting dashboard because of "Google free listings reporting API failed to respond with data".

Great point, I totally upvote it.

I'm removing the link as mentioned in https://github.com/woocommerce/google-listings-and-ads/pull/749#issuecomment-856868185 in favor of https://github.com/woocommerce/google-listings-and-ads/issues/768.

But I'd create another micro PR to change the event name for Dashboard.

tomalec commented on js/src/reports/utils.js:
```diff
@@ -90,24 +93,45 @@ export function aggregateIntervals( intervals1, intervals2 ) {
+) {
+	const metrics = paidTotals ? Object.keys( paidTotals ) : expectedFreeFields;
 	paidTotals = paidTotals || {};
-	return metrics.reduce(
-		( acc, key ) => ( {
+
+	return metrics.reduce( ( acc, key ) => {
+		let missingFreeListingsData = MISSING_FREE_LISTINGS_DATA.NONE;
+		let value = paidTotals[ key ];
+
+		if ( expectedFreeFields ) {
+			if ( ! expectedFreeFields.includes( key ) ) {
+				// We do not expect free to come.
+				missingFreeListingsData = MISSING_FREE_LISTINGS_DATA.FOR_METRIC;
+			} else if ( freeTotals[ key ] === undefined ) {
+				// We expected free data to come, but there is none.
+				missingFreeListingsData =
+					MISSING_FREE_LISTINGS_DATA.FOR_REQUEST;
+			} else {
+				// There is free listings data, sum with paid one.
+				value = ( paidTotals[ key ] || 0 ) + ( freeTotals[ key ] || 0 );
+			}
+		}
+
+		return {
 			...acc,
 			[ key ]: fieldsToPerformance(
-				( paidTotals[ key ] || 0 ) + ( freeTotals[ key ] || 0 ),
+				value,
 				undefined,
-				expectBoth && freeTotals[ key ] === undefined
+				missingFreeListingsData
 			),
-		} ),
-		{}
-	);
+		};
+	}, {} );
```
Yes, :confused: I totally agree. I tried to somewhat fix it, but I had no idea how to do it gently enough, to :
 - Make it actually simpler to read
 - Do not break any existing scenario (especially, that initially, I was trying to squeeze this PR to 1.0)
 
 WDYT of making a :scroll: follow-up issue, and clean it up once we have some automated tests to mimic major success and error scenarios of possible data coming from the API.

ecgan replied:
Well we can make a follow-up issue, but what can we do about the issue? How to clean it up? 😂  I have no idea now.

I think we can proceed with what we have here. It was just a feeling and thought that I have and I wanted to write it down. 😄

tomalec commented:
> Oh I see, tooltip appears on click. sweat_smile Cool, let's put the link in the tooltip then +1

Not really, it appears on mouseover. Then if you move your cursor slowly and gently enough, you can move it inside the tooltip, so it will not disappear.

> One last nitpick - Possible to change the link colour to #9ec2e6 instead? The blue there looks very dark against the grey tooltip background.

Sure thing. I changed alsto the other states 
	- regular and `:visited` to `#9ec2e6`,
	- `:hover` to 20% darker `#4d8fd1`
	-  `:active` to `25%` darker `#3982cc`

![Screencast of the link in all states.](https://user-images.githubusercontent.com/17435/121197783-e42dd280-c871-11eb-87e1-6c6eae4cb22b.gif)

I hope that's ok

-----


@j111q one more thing:

I made the _"merchants.google.com"_ link to actually point to `https://merchants.google.com/mc/reporting/dashboard` to take the user closer to their reports. 
I was thinking maybe we should 
- change the text to _"merchants.google.com/mc/reporting/dashboard"_, 
- make it redirect to precisely `https://merchants.google.com`
- or leave it as is _"merchants.google.com"_ -> `https://merchants.google.com/mc/reporting/dashboard`
?

tomalec commented:
It seems that making the link work correctly, is not so trivial with `@woocommerce/components.SummaryNumber` as when given `href` it becomes itself an anchor `<a>`. 
That itself is not very HTML valid, as it contains interactive elements ( delta with a tooltip), plus our interactive tooltip.

We could work around it by re-implementing `onClick` on top of `SummaryNumber`, but I think it may be overhead, and reaches too far from the OP.

Sorry for all the confusion made by me trying to add to much to the scope of this issue. I think we can still try to make this inner link work, but in a scope of a new PR, probably coupled with a PR to woocommerce-admin, to extend/adapt `SummaryNumber` to our needs.

In the meantime, I'll revert all the changes related to making the link clickable.

tomalec committed:
Update colors for links inside MetricNumber's tooltip.

Addresses https://github.com/woocommerce/google-listings-and-ads/pull/749#issuecomment-856415051

tomalec committed:
Remove the link from MetricNumber tooltip,

It's not trivial to make it work correctly, we could revisit it in https://github.com/woocommerce/google-listings-and-ads/issues/768.

tomalec committed:
Split `fieldsToPerformance` into two functions,

for readability, to address https://github.com/woocommerce/google-listings-and-ads/pull/749#discussion_r647295315

tomalec commented on js/src/reports/metric-number.js:
```diff
@@ -48,19 +49,43 @@ const MetricNumber = ( {
+		missingFreeListingsData === MISSING_FREE_LISTINGS_DATA.FOR_REQUEST
+	) {
+		infoText = (
+			<>
```
:construction: 
This breaks `aria-label` used below, as it renders `"aria-label="[object Object]"`

tomalec committed:
Avoid using `createInterpolateElement` in MetricNumber Tooltip,

Fix `aria-label` https://github.com/woocommerce/google-listings-and-ads/pull/749/files#r647604308

Separate the infos, apply gutters via CSS not `<br/>`

Addresses https://github.com/woocommerce/google-listings-and-ads/pull/749#discussion_r647314794

tomalec commented:
@ecgan I believe it's ready for another round of review :)

ecgan commented on js/src/reports/metric-number.js:
```diff
@@ -48,27 +49,50 @@ const MetricNumber = ( {
-			'This data is available for paid campaigns only.',
-			'google-listings-and-ads'
+	// And Free Listings API may not respond with data.
+	if ( missingFreeListingsData !== MISSING_FREE_LISTINGS_DATA.NONE ) {
+		infos.push(
+			__(
+				'This data is currently available for paid campaigns only.',
+				'google-listings-and-ads'
+			)
 		);
+	}
+	if ( missingFreeListingsData === MISSING_FREE_LISTINGS_DATA.FOR_REQUEST ) {
```
💅 Line 65 uses `!==`, and line 73 uses `===`. Sometimes I'm a bit worried about situation like this, where we might accidentally read it wrongly and thought both are `!==` or both are `===`. Personally, I _might_ write line 65 like the following: 

```js
// instead of: if ( missingFreeListingsData !== MISSING_FREE_LISTINGS_DATA.NONE ) {
if ( [ MISSING_FREE_LISTINGS_DATA.FOR_METRIC, MISSING_FREE_LISTINGS_DATA.FOR_REQUEST ].includes( missingFreeListingsData ) ) {
```

However, the above style is arguable too. Imagine that we have more MISSING_FREE_LISTINGS_DATA in the future, we might need to add them into the array, which can be problematic and easily missed too. 

Anyway, I guess it's probably fine to leave what you have here as is. 😄  Just voicing things out from my head.

ecgan reviewed (approved):
Second round of review, LGTM 👍

j111q commented:
> I made the "merchants.google.com" link to actually point to https://merchants.google.com/mc/reporting/dashboard to take the user closer to their reports.

I see, great that we can do a more specific link!

How about let's change it to:

`This data is currently available for paid campaigns only. 
Please try again later, or go to Google Merchant Center to track your performance for Google Free Listings.`

So like this:

![image](https://user-images.githubusercontent.com/65582817/121472269-89bc7f80-c9f3-11eb-969f-387d092e48f8.png)

and then we link directly to `https://merchants.google.com/mc/reporting/dashboard`

eason9487 committed:
Add Google MC link back to the tooltip and keep the string value for `aria-label`

Address https://github.com/woocommerce/google-listings-and-ads/issues/768

eason9487 committed:
Tweak tooltp style within the MetricNumber"

Address

- https://github.com/woocommerce/google-listings-and-ads/issues/768

And related comments

- https://github.com/woocommerce/google-listings-and-ads/pull/749#issuecomment-856415051
- https://github.com/woocommerce/google-listings-and-ads/pull/749#issuecomment-856794571
- https://github.com/woocommerce/google-listings-and-ads/pull/749#issuecomment-858328543

eason9487 committed:
Fix that the Google MC link is redirected locally

Address https://github.com/woocommerce/google-listings-and-ads/issues/768

eason9487 commented:
Found some ways to resolve the above issues (#768) and changed the link text. I pushed some commits about:

1. Tooltip link color and wording. And Its size was changed to `32em` for fitting the visual example at https://github.com/woocommerce/google-listings-and-ads/pull/749#issuecomment-858328543.
    ![image](https://user-images.githubusercontent.com/17420811/121508829-9d2e1180-ca18-11eb-8230-6ded07be3d2d.png)
2. Keep the pure text for the `aria-label` mentioned at https://github.com/woocommerce/google-listings-and-ads/pull/749#discussion_r647604308
    ![2021-06-10 15 58 06](https://user-images.githubusercontent.com/17420811/121509434-36f5be80-ca19-11eb-85ad-dce25146cd77.png)
3. Open a new tab to Google Merchant Center when clicking on link
    ![Kapture 2021-06-10 at 18 20 31](https://user-images.githubusercontent.com/17420811/121509745-8936df80-ca19-11eb-99f5-3be0dd4cd67b.gif)

eason9487 commented:
Hey @ecgan, please refer to this [diff link](https://github.com/woocommerce/google-listings-and-ads/pull/749/files/979e77b590c59a8f81e0f799e34c2fe8fc4e9486..d5d56b60becb9bd6d59ec7da0d92c0ebbd47304d) if it is convenient for you to look at the new changes again 🙏

ecgan commented on js/src/reports/metric-number.js:
```diff
@@ -59,24 +67,54 @@ const MetricNumber = ( {
 	}
 	if ( missingFreeListingsData === MISSING_FREE_LISTINGS_DATA.FOR_REQUEST ) {
+		const text = __(
+			'Please try again later, or go to <linkOfGoogleMerchantCenter /> to track your performance for Google Free Listings.',
```
At first, I was thinking why use `<linkOfGoogleMerchantCenter />` and not the more conventional way of `<link>Google Merchant Center</link>`, but after looking at the code below where we want to differentiate texts for UI and ARIA, I got to say this is pretty smart! 😃 👍 

💅  Very small nit: I'd name it as `googleMerchantCenterLink` to get rid of `"of"`. 😆

eason9487 replied:
Renamed by https://github.com/woocommerce/google-listings-and-ads/pull/749/commits/3b3fbda897e03261bc50d585949adc9c2a6003fb

ecgan commented on js/src/reports/metric-number.js:
```diff
@@ -59,24 +67,54 @@ const MetricNumber = ( {
+
+		// `aria-label` prop only accepts a pure text.
+		const textElement = createInterpolateElement( text, {
+			linkOfGoogleMerchantCenter: <>{ googleMCReportingDashboardURL }</>,
```
❔ I'm thinking, instead of just the URL, would it be better to set it like below, with the text "Google Merchant Center" and then the URL in parentheses? 🤔 

```suggestion
			linkOfGoogleMerchantCenter: (
				<>
					{ sprintf(
						// translators: %s: link to Google Merchant Center.
						__(
							'Google Merchant Center (%s)',
							'google-listings-and-ads'
						),
						googleMCReportingDashboardURL
					) }
				</>
			),
```

Note: need to import `sprintf` in the file.

ecgan reviewed (approved):
Tested locally, works okay. 👍 

Just one small thing to note: in the tooltip popup, if we click on the black popup, it would act like we are clicking the `<a>` element at the back and causes the `<SumaryNumber>` tab click / tab change. See gif below. 

![tooltip-click-tab-change](https://user-images.githubusercontent.com/417342/121541251-b85d4900-ca39-11eb-8c61-8aa837781602.gif)

eason9487 commented on js/src/reports/metric-number.js:
```diff
@@ -2,7 +2,11 @@
  * External dependencies
  */
 import { __ } from '@wordpress/i18n';
```
For batching commit with the suggested change #749 (comment)

```suggestion
import { __, sprintf } from '@wordpress/i18n';
```

eason9487 committed:
Add text "Google Merchant Center" with the URL for `aria-label`

Co-authored-by: Gan Eng Chin <ecgan@users.noreply.github.com>

eason9487 committed:
Rename linkOfGoogleMerchantCenter to googleMerchantCenterLink

Address https://github.com/woocommerce/google-listings-and-ads/pull/749#discussion_r649209001

eason9487 commented:
> Just one small thing to note: in the tooltip popup, if we click on the black popup, it would act like we are clicking the <a> element at the back and causes the <SumaryNumber> tab click / tab change.

That's an interesting catch! To address this, it needs to attach click event `e.stopPropagation(); e.preventDefault();` to `.components-popover__content` node within [Popover component](https://github.com/WordPress/gutenberg/blob/%40wordpress/components%4012.0.8/packages/components/src/popover/index.js#L618). But it doesn't expose prop to do so.

tomalec commented:
Sorry, for being late, but I posted some comments at https://github.com/woocommerce/google-listings-and-ads/issues/806, briefly, we now have an invalid HTML - we nest anchors, and in a perfect `woo/components` world, we could approach ARIA easier.
